### PR TITLE
Feature/36 search page

### PR DESCRIPTION
**Pull request by davideastmond**
This PR implements the search page functionality on the front end. Users are able to search users, public and private threads and thread comments.

### CODE REVIEW

File: app/src/components/search/Search.tsx [hunk 1]
```diff
@@ -16,6 +24,7 @@ function Search({ className }: { className: string }) {
         type="text"
         value={query}
         onChange={(e) => setQuery(e.target.value)}
+        onKeyPress={(e) => handleEnterKeyPress(e)}
       />
```
Comment (lily-law) [inline app/src/components/search/Search.tsx:29]: Maybe we could add an onblur={setQuery} here

**Review by lily-law — CHANGES_REQUESTED** [review]
Nice work on this. Sub-components is a good idea, I could make use of this on a few of the components to de-clutter a little. 
Looking at this each page seems to have the same logic, something we could look at doing is moving that into a higher order component, basically we could wrap the components we want replaced on searching in the search component it's self: 
```
<Search setOnSearchSubmit={setOnSearchSubmit}>
  <ContentForExample />
  <Avatar />
  <Post />
</Search> 
```
Then move the search bits from pages into the Search component... 
```
  const [searchTriggered, setSearchTriggered] = useState(false)
  const [searchResultData, setSearchResultData] = useState<ISearchResults>()
  const onSearchSuccess = (data: any) => {
    setSearchResultData(data);
  }
  const onSearchError = (data: any) => {
    // Some error occurred
    console.log("Error", data)
  }
  const onSearchSubmit = (queryString: string) => {
    setSearchTriggered(!!queryString)
    doSearch({ queryString: queryString, onSuccess: onSearchSuccess, onError: onSearchError })
  }
```
And add a prop and useEffect to set the onSearchSubmit so the pages can pass this to the TopBar (temporary solution, once we've refactored in redux this can be more direct).
The Search component can then just return back props.children if !searchTriggered or return the results.

File: app/.gitignore [hunk 2]
```diff
@@ -24,3 +24,5 @@
 npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
+
+/.vscode
```
Comment (lily-law) [inline app/.gitignore]: Maybe this is a typo, should it be `.vscode/` to prevent .vscode/settings.json from getting added?

**Review by lily-law — APPROVED** [review]
This is looking good :+1:

File: app/.gitignore (hunk 2 above)
Comment (davideastmond) [reply app/.gitignore]: We have two .vscode folders in this repo - at the root level and app level. One of them (I forgot which) was being included in my commits. This seemed to solve it. I will delete it and put it in the root level .gitignore ^_^

**Pull request merged**